rpc/internal/collector.service_test.go

```diff
@@ -0,0 +1,449 @@
+			id, group, getErr := svc.coordinator.getStream(test.collectorName.Name)
+			if test.hasErr {
+				assert.Error(t, registerErr)
+				assert.Error(t, getErr)
+
+			} else {
+				assert.NoError(t, registerErr)
+				assert.NotEmpty(t, id)
+				assert.NotNil(t, group)
+				assert.NoError(t, getErr)
+			}
+
+		})
+	}
+}
+
+func TestStreamEvent(t *testing.T) {
+	tmpDir, err := ioutil.TempDir(".", "stream-event-test")
+	require.NoError(t, err)
+	defer func() {
+		assert.NoError(t, os.RemoveAll(tmpDir))
+	}()
+	svc := getTestCollectorService(tmpDir)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	address := fmt.Sprintf("localhost:%d", 7070)
+	require.NoError(t, startCollectorService(ctx, svc, address))
+	client, err := getCollectorGRPCClient(ctx, address, []grpc.DialOption{grpc.WithInsecure()})
+	require.NoError(t, err)
+
+	for _, test := range []struct {
+		name     string
+		events   []*EventMetrics
+		resp     *PoplarResponse
+		register bool
+		hasErr   bool
+	}{
+		{
+			name:   "Unregistered",
+			events: []*EventMetrics{{Name: "DNE"}},
+			resp:   nil,
+			hasErr: true,
+		},
+		{
+			name:   "NoName",
+			events: []*EventMetrics{{}},
+			resp:   nil,
+			hasErr: true,
+		},
+		{
+			name: "DifferentNames",
+			events: []*EventMetrics{
+				{
+					Name: "collector",
+					Time: &timestamp.Timestamp{},
+					Timers: &EventMetricsTimers{
+						Total:    &duration.Duration{},
+						Duration: &duration.Duration{},
+					},
+				},
+				{
+					Name: "anotherName",
+					Time: &timestamp.Timestamp{},
+					Timers: &EventMetricsTimers{
+						Total:    &duration.Duration{},
+						Duration: &duration.Duration{},
+					},
+				},
+			},
+			resp:     nil,
+			register: true,
+			hasErr:   true,
+		},
+		{
+			name: "AddEvents",
+			events: []*EventMetrics{
+				{
+					Name: "collector",
+					Time: &timestamp.Timestamp{},
+					Timers: &EventMetricsTimers{
+						Total:    &duration.Duration{},
+						Duration: &duration.Duration{},
+					},
+				},
+				{
+					Name: "collector",
+					Time: &timestamp.Timestamp{},
+					Timers: &EventMetricsTimers{
+						Total:    &duration.Duration{},
+						Duration: &duration.Duration{},
+					},
+				},
+			},
+			resp:     &PoplarResponse{Name: "collector", Status: true},
+			register: true,
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			stream, err := client.StreamEvents(ctx)
+			require.NoError(t, err)
+
+			if test.register {
+				_, err = client.RegisterStream(ctx, &CollectorName{Name: "collector"})
+				require.NoError(t, err)
+			}
+
+			catcher := grip.NewBasicCatcher()
+			for i := 0; i < len(test.events); i++ {
+				catcher.Add(stream.Send(test.events[i]))
+			}
+			resp, err := stream.CloseAndRecv()
+			catcher.Add(err)
+			assert.Equal(t, test.resp, resp)
+
+			if test.hasErr {
+				assert.Error(t, catcher.Resolve())
+			} else {
+				assert.NoError(t, catcher.Resolve())
+			}
+		})
+	}
+
+	t.Run("MultipleStreams", func(t *testing.T) {
```

Comment:
In most cases I think this test will stream things in order just because `time.Now()` is monotonically increasing. Since there are some concurrency concerns, can we add a test for other cases like:
* Streams send different amounts of data points (e.g. stream 1 sends 1 data point, stream 2 sends 2 data points)
* Timestamp ordering guarantee (e.g. stream 1 sends time = 1, stream 2 sends time = 3, stream 1 sends time = 2, stream 2 sends time = 4)?

Reply:
The second point is tested here already, I will have the first stream add an extra data point.